Reduce image size built with Dockerfile

ryncsn opened the pull request:
Have a better split between the build stage and runtime stage.
Dropped gcc package, .git repo, and some caches in final image.
Reduced the image size by about 300M, also improve security.

Also speed up building by install most packegs in one place so
no need to fetch metadate many times.

SuperSandro2000 commented on Dockerfile:
```diff
@@ -43,56 +48,54 @@ RUN apt-get update && \
 	ln -s /opt/mastodon /mastodon && \
-	gem install bundler && \
+	echo "mastodon:$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 24 | mkpasswd -s -m sha-256)" | chpasswd && \
+	apt-get clean && \
```
```suggestion
```

Run automatically by docker.

SuperSandro2000 commented on Dockerfile:
```diff
@@ -3,6 +3,19 @@ FROM ubuntu:20.04 as build-dep
+	libncurses5-dev libffi-dev zlib1g-dev libssl-dev && \
+	apt-get clean
+
+ENV PATH="${PATH}:/opt/ruby/bin:/opt/node/bin:/opt/mastodon/bin"
```
```suggestion
ENV PATH="/opt/ruby/bin:/opt/node/bin:/opt/mastodon/bin:$PATH"
```

SuperSandro2000 commented on Dockerfile:
```diff
@@ -3,6 +3,19 @@ FROM ubuntu:20.04 as build-dep
+	libyaml-dev libgdbm-dev libreadline-dev libjemalloc-dev \
+	libicu-dev libidn11-dev libpq-dev \
+	libprotobuf-dev protobuf-compiler shared-mime-info \
+	libncurses5-dev libffi-dev zlib1g-dev libssl-dev && \
+	apt-get clean
```
```suggestion
	libncurses5-dev libffi-dev zlib1g-dev libssl-dev
```

ryncsn commented:
Thanks for the review, updated.

decentral1se reviewed (approved):
LGTM! Great to reduce the size of the image :rocket:

SuperSandro2000 commented on Dockerfile:
```diff
@@ -3,6 +3,18 @@ FROM ubuntu:20.04 as build-dep
+	libyaml-dev libgdbm-dev libreadline-dev libjemalloc-dev \
+	libicu-dev libidn11-dev libpq-dev \
+	libprotobuf-dev protobuf-compiler shared-mime-info \
+	libncurses5-dev libffi-dev zlib1g-dev libssl-dev
```
```suggestion
	libncurses5-dev libffi-dev zlib1g-dev libssl-dev \
	&& rm -rf /var/lib/apt/lists/*
```

ryncsn replied:
Good suggestion, I'll update it.

ryncsn commented:
Updated, rebased to latest main branch.

exstral reviewed (commented):
I've been testing some of these changes out in my other pull request #16211 so thought I would offer some thoughts.

It is a really nice find with removing gcc and moving the precompile step into the build-dep stage. In my testing I found these two changes really makes the image much smaller.

I'm less certain about putting all the apt installs together. I like having them separate to group software by usage. For example the packages required for building ruby go in one list, the packages for node in another, etc. From what I can see this does not affect the final image size, but I do like the speed improvement it brings :)

In my branch I've instead used the new fancy `--mount=type=cache` [Docker BuildKit feature](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md) to make all apt installs faster. But that is a bit more complex and it does bump the required docker version to 18.09.

I'm looking forward to this change being merged and rebasing my local docker setup on top of it :)

ryncsn commented:
Thanks for all the feedbacks, is there any other update I need to do? Any comments?

ryncsn commented:
Hi, just did a rebase, any comments?

ryncsn commented:
Hi, I just did another rebase.

shleeable commented:
I need to spend a little time and test this, but in theory I like it.

ryncsn commented:
Thank you for looking into this, I see there are conflicts again, I'll do a rebase this weekend, also do some test again.

ryncsn committed:
Dockerfile: install python3 explicitly

Dockerfile build may fail with:

gyp ERR! find Python Python is not set from command line or npm configuration
gyp ERR! find Python Python is not set from environment variable PYTHON
gyp ERR! find Python checking if \"python3\" can be used
gyp ERR! find Python - \"python3\" is not in PATH or produced an error
gyp ERR! find Python checking if \"python\" can be used
gyp ERR! find Python - executable path is \"/usr/bin/python\"
gyp ERR! find Python - version is \"2.7.18\"
gyp ERR! find Python - version is 2.7.18 - should be >=3.6.0
gyp ERR! find Python - THIS VERSION OF PYTHON IS NOT SUPPORTED

So install python3 explicitly.

Signed-off-by: Kairui Song <ryncsn@gmail.com>

ClearlyClaire commented:
I'm not very knowledgeable regarding Docker, but it seems ok overall. Some thoughts/questions so far:
- is there any benefit in moving the `assets:precompile` stage to the build-dep step? (I don't think it's a bad thing, I can't see a downside either, I'm just not sure how it's functionally different)
- what is your “improve security” claim about?
- I'm not sure why `gcc` was included in the final build step, I don't think it was ever needed, so I definitely agree with removing it
- I cannot find the place where you drop the `.git` repo; also I'm not sure how I feel about such a change, I guess it is not useful in most cases, but could be in some cases, especially if running on `edge` or custom images

SuperSandro2000 commented:
> * is there any benefit in moving the `assets:precompile` stage to the build-dep step? (I don't think it's a bad thing, I can't see a downside either, I'm just not sure how it's functionally different)

I think it boils down to what you want to cache. Recompiling the assets takes a while and when hacking the container this should increase the reused cache. For release creation in CI this should change nothing.

> I cannot find the place where you drop the .git repo; also I'm not sure how I feel about such a change, I guess it is not useful in most cases, but could be in some cases, especially if running on edge or custom images

It should be in the .dockerignore file. It should never be added to the container via ADD or COPY because as soon as you change anything about your repo you will invalidate your cache even if all the actual files stay the same. This can be as simple as fetching remote objects. Also it copies the entire git repository into the container which can be very big.

ClearlyClaire commented:
> > I cannot find the place where you drop the .git repo; also I'm not sure how I feel about such a change, I guess it is not useful in most cases, but could be in some cases, especially if running on edge or custom images
> 
> It should be in the .dockerignore file. It should never be added to the container via ADD or COPY because as soon as you change anything about your repo you will invalidate your cache even if all the actual files stay the same. This can be as simple as fetching remote objects. Also it copies the entire git repository into the container which can be very big.

Oh, right. Seems like it has been taken care of in #16282 already, so this isn't changed by this PR anymore.

ClearlyClaire commented:
> gcc was required for the `gem install bundler` command, and `assets:precompile` requires `gem install bundler`, so to remove gcc in the final stage, we have to move everything to the build-dep stage.

`bundler` is a runtime dependency though. If it is not included in the final image, that is going to be an issue.

> And you are right, "improve security" is really not very accurate here, I was thinking with one less package like `gcc`, there will be less attack face, I can remove that from the commit message if you think that's a false claim.
> 
> Thanks for the reply :)

ryncsn commented:
> > gcc was required for the `gem install bundler` command, and `assets:precompile` requires `gem install bundler`, so to remove gcc in the final stage, we have to move everything to the build-dep stage.
> 
> `bundler` is a runtime dependency though. If it is not included in the final image, that is going to be an issue.

`bundler` is still there, we just copied it from the build-dep stage, just the dependency needed for build `bundler` is not required and left in the build-dep stage.

There might be problem if one try to update/install gem in a docker image, but I don't think that how the image should be used, to update/install new gem, either install gcc again or, more properly, rebuild the docker image.

> 
> > And you are right, "improve security" is really not very accurate here, I was thinking with one less package like `gcc`, there will be less attack face, I can remove that from the commit message if you think that's a false claim.
> 
> I think that would indeed be better!

Updated.

ClearlyClaire commented:
> `bundler` is still there, we just copied it from the build-dep stage, just the dependency needed for build `bundler` is not required and left in the build-dep stage.

I cannot find where it actually gets copied from the build-dep stage though

ryncsn commented:
> > `bundler` is still there, we just copied it from the build-dep stage, just the dependency needed for build `bundler` is not required and left in the build-dep stage.
> 
> I cannot find where it actually gets copied from the build-dep stage though

There:
```
COPY --from=build-dep /opt/node /opt/node
COPY --from=build-dep /opt/ruby /opt/ruby
COPY --from=build-dep --chown=mastodon:mastodon /opt/mastodon /opt/mastodon
```

And also yarn may also bring-in something that need gcc to compile, I forgot to mention that in previous comment, yarn is called by `assets:precompile` . But anyway there are all in the build-dep stage and copied to the final stage.

ryncsn commented:
And just ran some command in the docker image and here is the result, everything seems working well:

```
mastodon@mastodon-sidekiq-78fc77ff7f-qtwvc:~$ which bundle
/opt/ruby/bin/bundle
mastodon@mastodon-sidekiq-78fc77ff7f-qtwvc:~$ which bundler
/opt/ruby/bin/bundler
mastodon@mastodon-sidekiq-78fc77ff7f-qtwvc:~$ ./bin/update
== Installing dependencies == 
The Gemfile's dependencies are satisfied 
yarn install v1.22.17 
[1/6] Validating package.json...         
[2/6] Resolving packages...
success Already up-to-date. 
Done in 3.01s.
                                                                 
== Updating database ==
== Removing old logs and tempfiles ==
== Restarting application server ==

mastodon@mastodon-sidekiq-78fc77ff7f-qtwvc:~$ ./bin/tootctl cache clear
OK

```

ClearlyClaire reviewed (dismissed):
Looks fine to me, although I'd prefer if someone more knowledgeable about Docker had a look too.

ryncsn commented:
@shleeable gently ping, can you help have a look?

shleeable commented:
https://github.com/mastodon/mastodon/blob/1c48090d47cdfec765ac4f701bb9ed654c6711bb/Dockerfile#L66

Is this redundant?

shleeable commented:
Having two copies is redundant?
![image](https://user-images.githubusercontent.com/17537000/158719717-5b83a4a8-4c4c-4457-9cf6-c7fd1a59d2cb.png)
This is untested, but this should work right?
![image](https://user-images.githubusercontent.com/17537000/158719891-b4af6d05-e868-4601-92c9-ba39b0796ad4.png)

shleeable commented:
```
FROM ubuntu:20.04 as build-dep

# Use bash for the shell
SHELL ["/bin/bash", "-c"]

# Install build stage deps
RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
RUN echo "Etc/UTC" > /etc/localtime && \
	apt-get update && \
	apt-get install -y --no-install-recommends \
    # Generic OS
    ca-certificates \
    wget \
    python3 \
    apt-utils \
    # Generic Build
    git \
    build-essential \
	bison \
    # Ruby
	libyaml-dev \
    libgdbm-dev \
    libreadline-dev \
    libjemalloc-dev \
	libicu-dev \
    libidn11-dev \
    libpq-dev \
	shared-mime-info \
	libncurses5-dev \
    libffi-dev \
    zlib1g-dev \
    libssl-dev && \
	rm -rf /var/lib/apt/lists/*

ENV PATH="/opt/ruby/bin:/opt/node/bin:/opt/mastodon/bin:${PATH}"

# Install Node v16 (LTS)
ENV NODE_VER="16.13.2"
RUN ARCH= && \
    dpkgArch="$(dpkg --print-architecture)" && \
  case "${dpkgArch##*-}" in \
    amd64) ARCH='x64';; \
    ppc64el) ARCH='ppc64le';; \
    s390x) ARCH='s390x';; \
    arm64) ARCH='arm64';; \
    armhf) ARCH='armv7l';; \
    i386) ARCH='x86';; \
    *) echo "unsupported architecture"; exit 1 ;; \
  esac && \
	cd ~ && \
	wget -q https://nodejs.org/download/release/v$NODE_VER/node-v$NODE_VER-linux-$ARCH.tar.gz && \
	tar xf node-v$NODE_VER-linux-$ARCH.tar.gz && \
	rm node-v$NODE_VER-linux-$ARCH.tar.gz && \
	mv node-v$NODE_VER-linux-$ARCH /opt/node

# Install Ruby 3.0
ENV RUBY_VER="3.0.3"
RUN cd ~ && \
	wget https://cache.ruby-lang.org/pub/ruby/${RUBY_VER%.*}/ruby-$RUBY_VER.tar.gz && \
	tar xf ruby-$RUBY_VER.tar.gz && \
	cd ruby-$RUBY_VER && \
	./configure --prefix=/opt/ruby \
	  --with-jemalloc \
	  --with-shared \
	  --disable-install-doc && \
	make -j"$(nproc)" > /dev/null && \
	make install && \
	rm -rf ../ruby-$RUBY_VER.tar.gz ../ruby-$RUBY_VER

RUN npm install -g npm@latest && \
	npm install -g yarn && \
	gem install bundler

COPY . /opt/mastodon/

RUN cd /opt/mastodon && \
	bundle config set --local deployment 'true' && \
	bundle config set --local without 'development test' && \
	bundle config set silence_root_warning true && \
	bundle install -j"$(nproc)" && \
	yarn install --pure-lockfile && \
	RAILS_ENV=production NODE_ENV="production" \
	OTP_SECRET=precompile_placeholder SECRET_KEY_BASE=precompile_placeholder \
	rails assets:precompile && \
	yarn cache clean && \
	rm -rf tmp

# Build runtime stage
FROM ubuntu:20.04

# Add more PATHs to the PATH
ENV PATH="/opt/ruby/bin:/opt/node/bin:/opt/mastodon/bin:${PATH}"

# Use bash for the shell
SHELL ["/bin/bash", "-o", "pipefail", "-c"]

# Create the mastodon user
ARG UID=991
ARG GID=991

# Install mastodon runtime deps
RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
RUN echo "Etc/UTC" > /etc/localtime && \
	apt-get update && \
	apt-get install -y --no-install-recommends \
    # Generic OS
    whois \
    wget \
    tini \
    apt-utils \
    file \
    ca-certificates \
    tzdata \
    # Ruby
	libssl1.1 \
    libpq5 \
    imagemagick \
    ffmpeg \
    libjemalloc2 \
	libicu66 \
    libprotobuf17 \
    libidn11 \
    libyaml-0-2 \
    libreadline8 && \
	rm -rf /var/cache && \
	rm -rf /var/lib/apt/lists/* && \
	addgroup --gid $GID mastodon && \
	useradd -m -u $UID -g $GID -d /opt/mastodon mastodon && \
	ln -s /opt/mastodon /mastodon && \
	echo "mastodon:$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 24 | mkpasswd -s -m sha-256)" | chpasswd

# Copy over mastodon source, and dependencies from building, and set permissions
COPY --from=build-dep /opt/node /opt/node
COPY --from=build-dep /opt/ruby /opt/ruby
COPY --from=build-dep --chown=mastodon:mastodon /opt/mastodon /opt/mastodon

# Run mastodon services in prod mode
ENV RAILS_ENV="production"
ENV NODE_ENV="production"

# Tell rails to serve static files
ENV RAILS_SERVE_STATIC_FILES="true"
ENV BIND="0.0.0.0"

# Set the run user
USER mastodon

# Set the work dir and the container entry point
WORKDIR /opt/mastodon
ENTRYPOINT ["/usr/bin/tini", "--"]
EXPOSE 3000 4000
```

shleeable commented:
```
FROM ubuntu:20.04 as build-dep

# Use bash for the shell
SHELL ["/bin/bash", "-c"]

# Install build stage deps
RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
RUN echo "Etc/UTC" > /etc/localtime && \
    apt-get update && \
    apt-get install -y --no-install-recommends \
    # Generic OS
    ca-certificates \
    wget \
    python3 \
    apt-utils \
    # Generic Build
    git \
    build-essential \
    bison \
    # Ruby
    libyaml-dev \
    libgdbm-dev \
    libreadline-dev \
    libjemalloc-dev \
    libicu-dev \
    libidn11-dev \
    libpq-dev \
    shared-mime-info \
    libncurses5-dev \
    libffi-dev \
    zlib1g-dev \
    libssl-dev && \
    rm -rf /var/lib/apt/lists/*

ENV PATH="/opt/ruby/bin:/opt/node/bin:/opt/mastodon/bin:${PATH}"

# Install Node v16 (LTS)
ENV NODE_VER="16.13.2"
RUN ARCH= && \
    dpkgArch="$(dpkg --print-architecture)" && \
  case "${dpkgArch##*-}" in \
    amd64) ARCH='x64';; \
    ppc64el) ARCH='ppc64le';; \
    s390x) ARCH='s390x';; \
    arm64) ARCH='arm64';; \
    armhf) ARCH='armv7l';; \
    i386) ARCH='x86';; \
    *) echo "unsupported architecture"; exit 1 ;; \
  esac && \
    cd ~ && \
    wget -q https://nodejs.org/download/release/v$NODE_VER/node-v$NODE_VER-linux-$ARCH.tar.gz && \
    tar xf node-v$NODE_VER-linux-$ARCH.tar.gz && \
    rm node-v$NODE_VER-linux-$ARCH.tar.gz && \
    mv node-v$NODE_VER-linux-$ARCH /opt/node

# Install Ruby 3.0
ENV RUBY_VER="3.0.3"
RUN cd ~ && \
    wget https://cache.ruby-lang.org/pub/ruby/${RUBY_VER%.*}/ruby-$RUBY_VER.tar.gz && \
    tar xf ruby-$RUBY_VER.tar.gz && \
    cd ruby-$RUBY_VER && \
    ./configure --prefix=/opt/ruby \
      --with-jemalloc \
      --with-shared \
      --disable-install-doc && \
    make -j"$(nproc)" > /dev/null && \
    make install && \
    rm -rf ../ruby-$RUBY_VER.tar.gz ../ruby-$RUBY_VER

RUN npm install -g npm@latest && \
    npm install -g yarn && \
    gem install bundler

COPY . /opt/mastodon/

RUN cd /opt/mastodon && \
    bundle config set --local deployment 'true' && \
    bundle config set --local without 'development test' && \
    bundle config set silence_root_warning true && \
    bundle install -j"$(nproc)" && \
    yarn install --pure-lockfile && \
    RAILS_ENV=production NODE_ENV="production" \
    OTP_SECRET=precompile_placeholder SECRET_KEY_BASE=precompile_placeholder \
    rails assets:precompile && \
    yarn cache clean && \
    rm -rf tmp

# Build runtime stage
FROM ubuntu:20.04

# Add more PATHs to the PATH
ENV PATH="/opt/ruby/bin:/opt/node/bin:/opt/mastodon/bin:${PATH}"

# Use bash for the shell
SHELL ["/bin/bash", "-o", "pipefail", "-c"]

# Create the mastodon user
ARG UID=991
ARG GID=991

# Install mastodon runtime deps
RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
RUN echo "Etc/UTC" > /etc/localtime && \
    apt-get update && \
    apt-get install -y --no-install-recommends \
    # Generic OS
    whois \
    wget \
    tini \
    apt-utils \
    file \
    ca-certificates \
    tzdata \
    # Ruby
    libssl1.1 \
    libpq5 \
    imagemagick \
    ffmpeg \
    libjemalloc2 \
    libicu66 \
    libprotobuf17 \
    libidn11 \
    libyaml-0-2 \
    libreadline8 && \
    rm -rf /var/cache && \
    rm -rf /var/lib/apt/lists/* && \
    addgroup --gid $GID mastodon && \
    useradd -m -u $UID -g $GID -d /opt/mastodon mastodon && \
    ln -s /opt/mastodon /mastodon && \
    echo "mastodon:$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 24 | mkpasswd -s -m sha-256)" | chpasswd

# Copy over mastodon source, and dependencies from building, and set permissions
COPY --from=build-dep /opt/node /opt/node
COPY --from=build-dep /opt/ruby /opt/ruby
COPY --from=build-dep --chown=mastodon:mastodon /opt/mastodon /opt/mastodon

# Run mastodon services in prod mode
ENV RAILS_ENV="production"
ENV NODE_ENV="production"

# Tell rails to serve static files
ENV RAILS_SERVE_STATIC_FILES="true"
ENV BIND="0.0.0.0"

# Set the run user
USER mastodon

# Set the work dir and the container entry point
WORKDIR /opt/mastodon
ENTRYPOINT ["/usr/bin/tini", "--"]
EXPOSE 3000 4000
```

ryncsn commented:
> Having two copies is redundant?
> ![image](https://user-images.githubusercontent.com/17537000/158719717-5b83a4a8-4c4c-4457-9cf6-c7fd1a59d2cb.png)
> This is untested, but this should work right?
> ![image](https://user-images.githubusercontent.com/17537000/158719891-b4af6d05-e868-4601-92c9-ba39b0796ad4.png)

Yes, that will also work. I wanted to keep two layers since can help reuse the lower layer, when only mastodon source code is changed, only the upper layer is rebuilt, and no need to reinstall yarn/bundle files. might be helpful for development. And since these two layers are all in the build stage so should have no size effect on the final result.

It's ok either way to me.

shleeable commented:
ooh that makes sense to me... might as well keep the dependencies and the source separated for visibility.

ok. All things considered this PR looks good to me.

shleeable commented:
https://blog.vamc19.dev/posts/dockerfile-copy-chmod/

ryncsn commented:
> https://blog.vamc19.dev/posts/dockerfile-copy-chmod/

Nice article, let me check if this PR can be optimized more.

ryncsn committed:
Reduce image size built with Dockerfile

Have a better split between the build stage and runtime stage.
Dropped gcc package, and some caches in final image.
Reduced the image size by about 300M.

Also speed up building by install most packegs in one place so
no need to fetch metadate many times.

Signed-off-by: Kairui Song <ryncsn@gmail.com>

ryncsn committed:
Dockerfile: split package names by newlines

This make it easier to read and maintain as suggested by Shlee,
sorted in alphanum order.

Signed-off-by: Kairui Song <ryncsn@gmail.com>

ryncsn commented:
Hi

I think this Dockerfile looks good and there is no issue caused by metadata change.

If I build it and check the intermediate layers: 
```
# podman build -t mastodon .
[2/2] STEP 6/18: RUN echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
--> Using cache 54b732654e640146c7a77c9536b901966adec5c884b412b152728036a4bdef52
--> 54b732654e6
[2/2] STEP 7/18: RUN echo "Etc/UTC" > /etc/localtime &&         apt-get update &&       apt-get install -y --no-install-recommends  apt-utils
 ca-certificates         file    tini    tzdata  wget        whois   ffmpeg  imagemagick     libicu66        libidn11        libjemalloc2    libpq5
  libprotobuf17       libreadline8    libssl1.1       libyaml-0-2     &&      addgroup --gid $GID mastodon &&     useradd -m -u $UID -g $GID -d /op
t/mastodon mastodon &&         ln -s /opt/mastodon /mastodon &&    echo "mastodon:$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 24 | mkpasswd -s
 -m sha-256)" | chpasswd &&         rm -rf /var/cache &&    rm -rf /var/lib/apt/lists/*
Get:1 http://archive.ubuntu.com/ubuntu focal InRelease [265 kB]
... snip ...
--> 196859fbf2c
[2/2] STEP 8/18: COPY --from=build-dep /opt/node /opt/node
--> 019f59d5e4f
[2/2] STEP 9/18: COPY --from=build-dep /opt/ruby /opt/ruby
--> 1bc6f7dd383
[2/2] STEP 10/18: COPY --from=build-dep --chown=mastodon:mastodon /opt/mastodon /opt/mastodon
--> fcf6672dd31
[2/2] STEP 11/18: ENV RAILS_ENV="production"
--> 002c0fb7865
[2/2] STEP 12/18: ENV NODE_ENV="production"
--> a8ab10ae3af

```

And...
```
# podman image ls -a
REPOSITORY                TAG         IMAGE ID      CREATED            SIZE
localhost/mastodon        latest      2268413c3af0  About an hour ago  1.4 GB
<none>                    <none>      181f20e42ed0  About an hour ago  1.4 GB
<none>                    <none>      649da529cd26  About an hour ago  1.4 GB
<none>                    <none>      ac094df4019b  About an hour ago  1.4 GB
<none>                    <none>      3fb6c40c3aa2  About an hour ago  1.4 GB
<none>                    <none>      7e119ea76689  About an hour ago  1.4 GB
<none>                    <none>      a8ab10ae3af1  About an hour ago  1.4 GB
<none>                    <none>      002c0fb78651  About an hour ago  1.4 GB
<none>                    <none>      fcf6672dd316  About an hour ago  1.4 GB
<none>                    <none>      1bc6f7dd383f  About an hour ago  753 MB
<none>                    <none>      019f59d5e4ff  About an hour ago  552 MB
<none>                    <none>      196859fbf2c0  About an hour ago  450 MB
<none>                    <none>      54b732654e64  6 days ago         76.1 MB
```

You can see the five major size increasement in 54b732654e64 196859fbf2c0 019f59d5e4ff 1bc6f7dd383f fcf6672dd316, which correspons to "ubuntu base image", "apt package installation", "copy ruby deps", "copy node deps", "copy runtime bins". 

It's a large image but at this point still looks reasonable. We may reduce the size of bundle/node_modules/runtime bins if we want to shrink the size even more. Maybe in another PR.

shleeable commented:
Looks fine to me. Merge away @ClearlyClaire @Gargron

ryncsn commented:
@ClearlyClaire @Gargron @shleeable gently ping? This PR is clean and reviewed for many times, will you help merge it?

shleeable commented:
Still good to merge.

decentral1se commented:
I hope this will be merged. The image is very large and takes a really long time to build on small bandwidth networks or low powered machines. So much so that the workflow can last easily over an hour when hacking something together...

ineffyble commented:
#20933 has hopefully solved some of the build performance issues although there's hopefully more that can be done.

nileshtrivedi commented:
Hoping for this to get merged as an 18 month old PR because other newer PRs related to Docker improvements continue to get piled up:

- https://github.com/mastodon/mastodon/pull/21044
- https://github.com/mastodon/mastodon/pull/21059
- https://github.com/mastodon/mastodon/pull/21137
- https://github.com/mastodon/mastodon/pull/21143
- https://github.com/mastodon/mastodon/pull/21165
- https://github.com/mastodon/mastodon/pull/21234
